recbole/model/general_recommender/line.py

```diff
@@ -0,0 +1,176 @@
+            cur[iid].add(uid)
+        return cur
+
+    def sampler(self,key_ids):
+
+        key_ids = np.array(key_ids.cpu())
+        key_num = len(key_ids)
+        total_num = key_num
+        value_ids = np.zeros(total_num, dtype=np.int64)
+        check_list = np.arange(total_num)
+        key_ids = np.tile(key_ids, 1)
+        while len(check_list) > 0:
+            value_ids[check_list] = self.random_num(len(check_list))
+            check_list = np.array([i for i, used, v in
+                                   zip(check_list, self.used_ids[key_ids[check_list]], value_ids[check_list])
+                                   if v in used])
+
+        return torch.tensor(value_ids,device=self.device)
+
+    def random_num(self, num):
+        value_id = []
+        self.random_pr %= self.random_list_length
+        while True:
+            if self.random_pr + num <= self.random_list_length:
+                value_id.append(self.random_list[self.random_pr: self.random_pr + num])
+                self.random_pr += num
+                break
+            else:
+                value_id.append(self.random_list[self.random_pr:])
+                num -= self.random_list_length - self.random_pr
+                self.random_pr = 0
+                np.random.shuffle(self.random_list)
+        return np.concatenate(value_id)
+
+    def get_random_list(self):
+        return np.arange(1, self.n_users)
+
+    def forward(self,h,t):
+
+        h_embedding = self.user_embedding(h)
+        t_embedding = self.item_embedding(t)
+
+        return torch.sum(h_embedding.mul(t_embedding), dim=1)
+
+    def context_forward(self, h, t, field):
+
+        if field == "uu":
+            h_embedding = self.user_embedding(h)
+            t_embedding = self.item_context_embedding(t)
+        else:
+            h_embedding = self.item_embedding(h)
+            t_embedding = self.user_context_embedding(t)
+
+        return torch.sum(h_embedding.mul(t_embedding), dim=1)
+
+    def calculate_loss(self, interaction):
+
+        user = interaction[self.USER_ID]
+        pos_item = interaction[self.ITEM_ID]
+        neg_item = interaction[self.NEG_ITEM_ID]
+
+        score_pos = self.forward(user,pos_item)
+
+        # randomly train i-i relation and u-u relation with u-i relation
+        if random.random()<0.5:
+            score_neg = self.forward(user, neg_item)
+            score_pos_con = self.context_forward(user, pos_item, 'uu')
+            score_neg_con = self.context_forward(user, neg_item, 'uu')
+        else:
+            # sample negative user for item
+            neg_user = self.sampler(pos_item)
+            score_neg = self.forward(neg_user,pos_item)
+            score_pos_con = self.context_forward(pos_item, user,'ii')
+            score_neg_con = self.context_forward(pos_item, neg_user,'ii')
+
+        ones = torch.ones(len(score_pos),device=self.device)
```

Comment:
It can be generated in ``__init__`` to speed up.

Reply:
Variable ones has different length since the last batch's length is not equal to batch size. So I think it cannot be generated in __init__.

Reply:
Got it, thx!